File: vpn.js
#!/usr/bin/env node

"use strict";

var vpnc = require("vpnc");
var chalk = require("chalk");
var db = require("mssql");
var config = require("./config");
var log = console.log;

/* uncomment the next line to use sheldon.forestry.oregonstate.edu */
// config.db.server = "sheldon.forestry.oregonstate.edu";

function stringify(o) {
    if (o !== undefined && o !== null) {
        if (o.constructor === Object || o.constructor === Array) {
            return JSON.stringify(o, null, 2);
        } else {
            return o.toString();
        }
    }
}

var vpn = {

    doStuff: function () {
        var self = this;
    
        /* insert commands you want to carry out while connected to the vpn here */

        var conn = new db.Connection(config.db, function (err) {
            if (err) throw err;

            /* insert sql queries here! */

            var query = "SELECT TOP 1 * FROM metdat.dbo.phrsc_table2 ORDER BY tmstamp ASC";
            var request = conn.request();
            request.query(query, function (err, recordset) {
                if (err) throw err;
                log(chalk.cyan(stringify(recordset)));
                self.disconnect();
            });
        });

        conn.on("error", function (err) {
            if (err) log("Database error:", err);
        });
    },

    disconnect: function () {
        vpnc.disconnect(function (err, code) {
            if (err) {
                log("Error disconnecting VPN:", err);
            } else {
                log("VPN", chalk.yellow("disconnected [" + code + "]"));
                // uncomment next line to quit Node automatically on disconnect
                // process.exit(0);
            }
        });
    },

    connect: function () {
        var self = this;
        vpnc.connect(config.vpn, function (err, code) {
            if (err) {
                log("Error connecting VPN:", err);
            } else {
                try {
                    self.doStuff();
                } catch (ex) {
                    log("Couldn't do stuff. Are you connected to the VPN?");
                    self.disconnect();
                }
            }
        });
    }
};

vpnc.available(function (err, version) {
    if (err) {
        log("vpnc unavailable:", err);
    } else {
        log("Found vpnc:", chalk.gray(stringify(version)));
        vpn.connect();
    }
});

process.on("exit", function (code) { if (code) vpn.disconnect(); });

module.exports = vpn;

File: vpn.py
#!/usr/bin/env python
"""A post-interview present for Fox :)"""
import os
import json
import subprocess
import pymssql
from contextlib import contextmanager

HERE = os.path.dirname(os.path.realpath(__file__))

def do_stuff(cursor):

    ######################
    # YOUR LOGIC HERE :) #
    ######################

    query = "SELECT TOP 1 * FROM metdat.dbo.phrsc_table2 ORDER BY tmstamp ASC"
    cursor.execute(query)
    for row in cursor:
        print row

def vpn_connect(config):

    # create properly formatted vpnc config file
    tempfile = os.path.join(HERE, "tempvpnc.conf")
    with open(tempfile, "w+") as f:
        print >> f, (
            "IPSec gateway %(IPSec_gateway)s\n"
            "IPSec ID %(IPSec_ID)s\n"
            "IPSec secret %(IPSec_secret)s\n"
            "IKE Authmode %(IKE_Authmode)s\n"
            "Xauth username %(Xauth_username)s\n"
            "Xauth password %(Xauth_password)s"
        ) % config

    # move config file to /etc/vpnc
    subprocess.check_call(["sudo", "mv", tempfile, "/etc/vpnc/"])
    subprocess.check_call(["sudo", "chown", "root:root", "/etc/vpnc/tempvpnc.conf"])
    subprocess.check_call(["sudo", "chmod", "600", "/etc/vpnc/tempvpnc.conf"])

    # connect to vpnc
    subprocess.check_call(["sudo", "vpnc", "tempvpnc"], env=os.environ)    

def vpn_disconnect():

    # disconnect from vpnc
    subprocess.check_call(["sudo", "vpnc-disconnect"])

    # remove the generated config file
    subprocess.check_call(["sudo", "rm", "/etc/vpnc/tempvpnc.conf"])

@contextmanager
def vpnc(config):
    vpn_connect(config)
    yield
    vpn_disconnect()

def main():
    with open(os.path.join(HERE, "config.json")) as configfile:
        config = json.load(configfile)
    with vpnc(config["vpn"]):
        db_params = {
            "server": config["db"]["server"] + ":" + config["db"]["port"],
            "user": config["db"]["user"],
            "password": config["db"]["password"],
            "database": config["db"]["database"],
        }
        with pymssql.connect(**db_params) as conn:
            with conn.cursor() as cursor:
                do_stuff(cursor)

if __name__ == "__main__":
    main()
